gui: Showing Local Addresses in Node Window

jadijadi opened the pull request:
This change adds one line to the Node Window (debugwindow.ui)
under the Network section and shows the LocalAddresses there.

fixed#564

<!--
*** Please remove the following help text before submitting: ***

Pull requests without a rationale and clear improvement may be closed
immediately.

GUI-related pull requests should be opened against
https://github.com/bitcoin-core/gui
first. See CONTRIBUTING.md
-->

<!--
Please provide clear motivation for your patch and explain how it improves
Bitcoin Core user experience or Bitcoin Core developer experience
significantly:

* Any test improvements or new tests that improve coverage are always welcome.
* All other changes should have accompanying unit tests (see `src/test/`) or
  functional tests (see `test/`). Contributors should note which tests cover
  modified code. If no tests exist for a region of modified code, new tests
  should accompany the change.
* Bug fixes are most welcome when they come with steps to reproduce or an
  explanation of the potential issue as well as reasoning for the way the bug
  was fixed.
* Features are welcome, but might be rejected due to design or scope issues.
  If a feature is based on a lot of dependencies, contributors should first
  consider building the system outside of Bitcoin Core, if possible.
* Refactoring changes are only accepted if they are required for a feature or
  bug fix or otherwise improve developer experience significantly. For example,
  most "code style" refactoring changes require a thorough explanation why they
  are useful, what downsides they have and why they *significantly* improve
  developer experience or avoid serious programming bugs. Note that code style
  is often a subjective matter. Unless they are explicitly mentioned to be
  preferred in the [developer notes](/doc/developer-notes.md), stylistic code
  changes are usually rejected.
-->

<!--
Bitcoin Core has a thorough review process and even the most trivial change
needs to pass a lot of eyes and requires non-zero or even substantial time
effort to review. There is a huge lack of active reviewers on the project, so
patches often sit for a long time.
-->

jarolrod commented on src/qt/rpcconsole.cpp:
```diff
@@ -958,6 +959,18 @@ void RPCConsole::updateNetworkState()
     ui->numberOfConnections->setText(connections);
+
+    // localAddressess
+    std::string localAddresses="";
```
you could just make this a QString, then `item.first.ToString()` -> `QString::fromStdString(item.first.ToString())`

jarolrod replied:
Should rename from `localAddresses` -> `local_addresses` to keep in line with our coding style

jarolrod commented on src/qt/rpcconsole.cpp:
```diff
@@ -958,6 +959,18 @@ void RPCConsole::updateNetworkState()
+    std::string localAddresses="";
+    LOCK(g_maplocalhost_mutex);
+    for (const std::pair<const CNetAddr, LocalServiceInfo> &item : mapLocalHost)
+        localAddresses += item.first.ToString() + ":" + std::to_string(item.second.nPort) + ", ";
```
to fix the linter issue arrising from `std::to_string(item.second.nPort), and keeping in mind that `locale_addresses` is now a QString: 

```suggestion
        local_addresses += QString::fromStdString(item.first.ToString()) + ":" + QString::number(item.second.nPort) + ", ";
```

`QString::number` fixes the linter issue, from [Qt Docs](https://doc.qt.io/qt-5/qstring.html#number): "The formatting always uses [QLocale::C](https://doc.qt.io/qt-5/qlocale.html#Language-enum), i.e., English/UnitedStates"

jadijadi replied:
Thank you very much. Greatly appreciate. Will do the changes & will commit.

jarolrod reviewed (commented):
concept ack

full diff: 

``` diff
diff --git a/src/qt/rpcconsole.cpp b/src/qt/rpcconsole.cpp
index 01989b5a0..080f9b979 100644
--- a/src/qt/rpcconsole.cpp
+++ b/src/qt/rpcconsole.cpp
@@ -961,15 +961,16 @@ void RPCConsole::updateNetworkState()
     ui->numberOfConnections->setText(connections);

     // localAddressess
-    std::string localAddresses="";
+    QString local_addresses;
     LOCK(g_maplocalhost_mutex);
     for (const std::pair<const CNetAddr, LocalServiceInfo> &item : mapLocalHost)
-        localAddresses += item.first.ToString() + ":" + std::to_string(item.second.nPort) + ", ";
-    if (localAddresses.length() > 2)
-        localAddresses[localAddresses.length()-2] = 0;
+        local_addresses += QString::fromStdString(item.first.ToString()) + ":" + QString::number(item.second.nPort) + ", ";
+    if (local_addresses.length() > 2)
+        local_addresses[local_addresses.length()-2] = 0;
     else
-        localAddresses = "None";
-    ui->localAddresses->setText(QString::fromStdString(localAddresses));
+        //: Signals to the user that they do not have any local addresses.
+        local_addresses = tr("None");
+    ui->localAddresses->setText(local_addresses);

 }
```

furszy commented on src/qt/rpcconsole.cpp:
```diff
@@ -49,6 +49,7 @@
 
 #include <chrono>
 
+
```
Can remove this extra added space

jadijadi replied:
done. will push.

furszy reviewed (commented):
Aside from the code comments, there is an important topic to mention:

Let's not access fields from the net layer in a GUI widget/view, the sources architecture is widget -> model -> interface -> backend.

Better to create the proper flow to access the information that you are looking for:
rpcconsole -> clientModel -> node interface -> net layer.

In this way, the field access is encapsulated, layers responsibilities are respected and the same function can be used, and tested, from other user facing end points like the RPC server equally.

jarolrod commented on src/qt/rpcconsole.cpp:
```diff
@@ -958,6 +959,18 @@ void RPCConsole::updateNetworkState()
+    if (localAddresses.length() > 2)
+        localAddresses[localAddresses.length()-2] = 0;
+    else
+        localAddresses = "None";
```
This is a review on an earlier state of the PR, the following would be the appropriate diff to express your idea:

``` diff
diff --git a/src/qt/rpcconsole.cpp b/src/qt/rpcconsole.cpp
index 386b9373c..a249d7895 100644
--- a/src/qt/rpcconsole.cpp
+++ b/src/qt/rpcconsole.cpp
@@ -964,13 +964,12 @@ void RPCConsole::updateNetworkState()
     // localAddressess
     QString local_addresses;
     LOCK(g_maplocalhost_mutex);
-    for (const std::pair<const CNetAddr, LocalServiceInfo> &item : mapLocalHost)
-        local_addresses += QString::fromStdString(item.first.ToString()) + ":" + QString::number(item.second.nPort) + ", ";
-    if (local_addresses.length() > 2)
-        local_addresses[local_addresses.length()-2] = 0;
-    else
-        // Signals to the user that they do not have any local addresses.
-        local_addresses = tr("None");
+    for (auto it = mapLocalHost.begin(); it != mapLocalHost.end();) {
+        local_addresses += QString::fromStdString(it->first.ToString()) + ":" + QString::number(it->second.nPort) + ", ";
+        if (++it != mapLocalHost.end()) local_addresses += ", ";
+    }
+    //: Signals to the user that they do not have any local addresses.
+    if (local_addresses.isEmpty()) local_addresses = tr("None");
     ui->localAddresses->setText(local_addresses);
 }
```

furszy replied:
I pointed to the previous state on purpose. My main suggestion was to encapsulate the mutex and `mapLocalHost` access in the backend, which is Qt/GUI independent. So cannot use QString from there. A plain std::string that will need to be converted once it pass through the model class or directly in the widget.

furszy replied:
> Dear @furszy thanks for the valuable input. Applied most of them except this encapsulation part. I understand your rational and I think it's a better design. But since the number of connections are also calculated here, I did the same with Local Address.

The connections aren't calculated here (on the widget as you are doing for the external addresses). It's encapsulated inside the connections manager. The flow is the following one widget -> model -> node -> conman:

`updateNetworkState` calls `clientModel->getNumConnections` which calls the interface `node.getNodeCount`, which calls the connection manager.

jadijadi replied:
Thanks. Did the flow as you said. Now we are querying layer by layer. Please comment if it still needs improvement. I decided to pass the `std::map<CNetAddr, LocalServiceInfo>` to keep it flexible enough for different use cases; for example in this specific usecase (showing IPs in GUI), i was able to do check for "IsI2P" on CNetAddrs.

jadijadi commented on src/qt/forms/debugwindow.ui:
```diff
@@ -249,6 +249,29 @@
+          <cursorShape>IBeamCursor</cursorShape>
+         </property>
+         <property name="text">
+          <string>N/A</string>
```
thanks. fixed this. will push.

jadijadi commented on src/qt/rpcconsole.cpp:
```diff
@@ -949,6 +949,7 @@ void RPCConsole::message(int category, const QString &message, bool html)
 
 void RPCConsole::updateNetworkState()
 {
+    // numberOfConnections
```
You are right. Variable names are self explanatory. will remove.

shaavan commented on src/qt/forms/debugwindow.ui:
```diff
@@ -249,6 +249,29 @@
         </widget>
        </item>
        <item row="9" column="0">
+        <widget class="QLabel" name="label_14">
```
nit:

```suggestion
        <widget class="QLabel" name="labelLocalAddresses">
```

jadijadi replied:
thanks for the checks & ACK @shaavan . I kept the label_14 in accordance in most of other fields which had the same pattern.

regarding the listening port, you can test with something like (for testnet):

```
testnet=1
blocksonly=1
[test]
#prune=1000 # this has no effect during syncing
externalip=100.200.100.10
rpcbind=127.0.0.1
rpcallowip=0.0.0.0/0
rpcuser=user
rpcpassword=pass
```

as your bitcoin.conf to force a listening IP.

shaavan reviewed (commented):
Concept ACK 

- I agree with showing the **Local Address** on the **Information Tab**.
- I successfully compiled the PR on `Ubuntu 22.04` with `Qt version 5.15.3`.
- I was able to verify that the **Local Address** option is displayed correctly on the **Information** Tab.

**Screenshot:**

**![Screenshot from 2022-07-05 17-09-17](https://user-images.githubusercontent.com/85434418/177327836-0a0f920f-b46a-48ca-9e79-6ffbda17f4cf.png)**

Currently, I am figuring out how to add local addresses to check if they are displayed correctly. It would be awesome if someone could point towards documentation on how to do so.

shaavan reviewed (approved):
ACK c47f01bf25e1353b51c24b9b31260a539a394b1a

Continuing on my [previous review](https://github.com/bitcoin-core/gui/pull/626#pullrequestreview-1028560104):

I was able to add local addresses successfully and tested that those were displayed perfectly on the GUI under the **Local Address** option.

**Method of testing:**

I added the following options in the `bitcoin.conf` file.

```
[signet]
externalip=100.200.100.11
externalip=100.200.100.10
externalip=100.200.100.12
```

The IP addresses, along with the port number, appeared in the GUI.

**Screenshot:**
![Screenshot from 2022-07-09 19-36-14](https://user-images.githubusercontent.com/85434418/178109522-9b8f688a-1506-4c64-a1ee-ada6e36e9f40.png)

Note: The `IP:port` appears in the sorted order irrespective of the order provided in the `bitcoin.conf` file. This behavior is the same as the `-netinfo` RPC command.

**Screenshot:**
![image](https://user-images.githubusercontent.com/85434418/178109635-5759cc35-1433-4f6d-b790-caa45e67a7e9.png)

Thanks, @jadijadi, for the help on adding the listening ports!

jarolrod commented on src/qt/forms/debugwindow.ui:
```diff
@@ -249,6 +249,29 @@
        <item row="9" column="0">
+        <widget class="QLabel" name="label_14">
+         <property name="text">
+          <string>Local Address</string>
```
we could add a translator comment here by adding in the `extracomment` field:

```suggestion
          <string extracomment="nice translator comment">Local Address</string>
```

jarolrod replied:
While here, if there's more than one local address; I guess this should take the plural form, no?

kristapsk replied:
Or could be just changed to "Local Address(es)" maybe?

jarolrod reviewed (commented):
Code Review ACK c47f01bf25e1353b51c24b9b31260a539a394b1a

Some nits

pablomartin4btc reviewed (commented):
Concept ACK and tested c47f01bf25e1353b51c24b9b31260a539a394b1a

![image](https://github.com/bitcoin-core/gui/assets/110166421/68f0d076-50dc-447a-a4e2-786914068b69)

Thanks @furszy for emphasizing the importance of a good code design and architectural integrity. These principles are crucial for the long-term maintainability and scalability of our project. I totally agree with his code change recommendation in terms of the flow/ pattern to follow.

In terms of the UI itself, I'd use the plural form "Local Addresses", as @jarolrod also pointed it out, to be consistent with the `getnetworkinfo` rpc command output.

![image](https://github.com/bitcoin-core/gui/assets/110166421/8003120a-5bbe-46de-bc30-f0d63bef888c)

kristapsk commented:
Concept ACK, kinda works.

Minor nit noticed - could avoid adding ":0" as a port for I2P addresses.

kristapsk commented:
Another thing - list is currently comma separated, without wrapping. Wouldn't it be too long if somebody has IPv4, IPv6, DJDNS, Tor and I2P addresses at the same time?

hebasto commented:
This PR still have a couple of unaddressed comments: https://github.com/bitcoin-core/gui/pull/626#discussion_r915334998 and https://github.com/bitcoin-core/gui/pull/626#discussion_r1274077350.

@jadijadi Are you still working on this?

jadijadi commented:
> This PR still have a couple of unaddressed comments: [#626 (comment)](https://github.com/bitcoin-core/gui/pull/626#discussion_r915334998) and [#626 (comment)](https://github.com/bitcoin-core/gui/pull/626#discussion_r1274077350).
> 
> @jadijadi Are you still working on this?

Will check and comment soon

jadijadi commented:
> Concept ACK, kinda works.
> 
> Minor nit noticed - could avoid adding ":0" as a port for I2P addresses.

thanks for the suggestion, did this.

jadijadi commented:
> Another thing - list is currently comma separated, without wrapping. Wouldn't it be too long if somebody has IPv4, IPv6, CJDNS, Tor and I2P addresses at the same time?

Thanks for this point, it is fixed and the widget worswraps and extends when needed.

jadijadi commented on src/qt/rpcconsole.cpp:
```diff
@@ -958,6 +958,17 @@ void RPCConsole::updateNetworkState()
+    LOCK(g_maplocalhost_mutex);
+    for (auto it = mapLocalHost.begin(); it != mapLocalHost.end();) {
+        local_addresses += QString::fromStdString(it->first.ToString()) + ":" + QString::number(it->second.nPort);
+        if (++it != mapLocalHost.end()) local_addresses += ", ";
```
changed to a normal loop and checking for .begin().

jadijadi commented:
> This PR still have a couple of unaddressed comments: [#626 (comment)](https://github.com/bitcoin-core/gui/pull/626#discussion_r915334998) and [#626 (comment)](https://github.com/bitcoin-core/gui/pull/626#discussion_r1274077350).

Dear @hebasto , addressed all issues, updated the code & replied to queries. Thanks for the follow up.

furszy commented on src/qt/rpcconsole.cpp:
```diff
@@ -975,6 +975,20 @@ void RPCConsole::updateNetworkState()
     }
 
     ui->numberOfConnections->setText(connections);
+
+    QString local_addresses;
+    std::map<CNetAddr, LocalServiceInfo> mapHosts = clientModel->getLocalAddresses();
+    for (auto it = mapHosts.begin(); it != mapHosts.end(); it++) {
+        if (it != mapHosts.begin()) local_addresses += ", ";
+        if (it->first.IsI2P())
+            local_addresses += QString::fromStdString(it->first.ToStringAddr());
+        else
+            local_addresses += QString::fromStdString(it->first.ToStringAddr()) + ":" + QString::number(it->second.nPort);
+    }
```
Maybe?
```c++
QString local_addresses;
std::map<CNetAddr, LocalServiceInfo> hosts = clientModel->getLocalAddresses();
for (const auto& [addr, info] : hosts) {
    local_addresses += QString::fromStdString(it->first.ToStringAddr());
    if (!addr.IsPrivacyNet()) local_addresses += ":" + QString::number(it->second.nPort);
    local_addresses += ", ";
}
local_addresses.chop(2); // remove last ", "
```

jadijadi replied:
This is more beautiful and understandable. Replaced the `it->first` & `it->second` with `addr` & `info`:

```
    for (const auto& [addr, info] : hosts) {
        local_addresses += QString::fromStdString(addr.ToStringAddr());
        if (!addr.IsPrivacyNet()) local_addresses += ":" + QString::number(info.nPort);
        local_addresses += ", ";
    }
    local_addresses.chop(2); // remove last ", "
```

furszy reviewed (commented):
thanks for the encapsulation jadijadi.

jadijadi commented on src/node/interfaces.cpp:
```diff
@@ -281,6 +281,13 @@ class NodeImpl : public Node
     }
+    std::map<CNetAddr, LocalServiceInfo> getLocalAddresses() override
+    {
+        std::map<CNetAddr, LocalServiceInfo> empty;
+
+        return m_context->connman ? m_context->connman->getLocalAddresses() : empty;
```
Thanks for all the support and patient. Applying them. The `{}` was also my first try but I got:
`error: initializer list cannot be used on the right hand side of operator ':'`
Now I changed it to 
```
        if (m_context->connman)
            return m_context->connman->getLocalAddresses();
        else
            return {};
```
to prevent that ugly `empty` variable of mine.

luke-jr commented on src/qt/rpcconsole.cpp:
```diff
@@ -975,6 +975,18 @@ void RPCConsole::updateNetworkState()
+    std::map<CNetAddr, LocalServiceInfo> hosts = clientModel->getLocalAddresses();
+    for (const auto& [addr, info] : hosts) {
+        local_addresses += QString::fromStdString(addr.ToStringAddr());
+        if (!addr.IsPrivacyNet()) local_addresses += ":" + QString::number(info.nPort);
```
AFAIK Tor hidden services do still have port numbers

luke-jr commented:
Also, please split the interfaces changes into a different commit than the GUI changes.

luke-jr commented:
Might be a good improvement to cherry-pick [e4fddc9f395819d0bdf74003676aa563929756fa](https://github.com/luke-jr/bitcoin-core-gui/commit/e4fddc9f395819d0bdf74003676aa563929756fa) and use it to update; in testing, I found myself watching for None to change, but it never did because the string isn't updated.

jadijadi commented:
> Also, please split the interfaces changes into a different commit than the GUI changes.

did that. cleaner and atomic. thanks.

luke-jr commented:
>did that. cleaner and atomic. thanks.

Hmm, it's not that way in your new push tho

luke-jr commented:
Somehow you introduced a typo in the commit message >_<

jadijadi commented:
> Somehow you introduced a typo in the commit message >_<

Oh.. sorry. fixed and added a spellchecker.

jadijadi commented:
@luke-jr May I kindly ask for a review on this PR?

hebasto commented:
cc @furszy

furszy commented on src/qt/clientmodel.h:
```diff
@@ -13,6 +13,9 @@
 #include <sync.h>
 #include <uint256.h>
 
+#include <net.h>
```
Could remove the need for adding this dependency here: 
```diff
diff --git a/src/qt/clientmodel.h b/src/qt/clientmodel.h
--- a/src/qt/clientmodel.h	(revision ea576b65592c23895f80c9116f881ff063253ff8)
+++ b/src/qt/clientmodel.h	(date 1722363555212)
@@ -13,7 +13,6 @@
 #include <sync.h>
 #include <uint256.h>
 
-#include <net.h>
 #include <netaddress.h>
 
 class BanTableModel;
@@ -22,6 +21,7 @@
 class PeerTableModel;
 class PeerTableSortProxy;
 enum class SynchronizationState;
+struct LocalServiceInfo;
 
 namespace interfaces {
 class Handler;
```

This will avoid including the backend `net.h` file in all the gui files that are already including `clientmodel.h`.

Ideally, `net.h` structs/classes shouldn't be accesible by the GUI widgets.  But no prob, it seems we are already including it in `rpcconsole.h` too.

jadijadi replied:
Thanks @furszy , this looks better. Applied it and pushed the new version with not `net.h` and with the `struct` only; as you suggested.

luke-jr replied:
You changed the net.h stuff in the second commit, after adding it in the first

jadijadi replied:
thanks! rebased.

furszy commented:
utACK

Maybe add a QLabel hint describing what "local addresses" are for GUI users? (and/or use another term for them).

jadijadi commented:
> utACK
> 
> Maybe add a QLabel hint describing what "local addresses" are for GUI users? (and/or use another term for them).

Any suggestion? English is not my first language. Something like "Addresses where Bitcoin Core binds to"? On *DataDir* we have the tooltip on the value column. I'll also keep it like that.

pablomartin4btc commented:
> > Maybe add a QLabel hint describing what "local addresses" are for GUI users? (and/or use another term for them).
> 
> Any suggestion? English is not my first language. Something like "Addresses where Bitcoin Core binds to"? On _DataDir_ we have the tooltip on the value column. I'll also keep it like that.

Perhaps any of these (or a combination):

- Each local address broadcast by this node.

- IP addresses of your local machine or network interface that the node uses to connect to the Bitcoin network

- Network addresses that your Bitcoin node is currently using to communicate with other nodes.

pablomartin4btc reviewed (commented):
<details>
<summary>re tACK bb5bb2001cf805301f56aa1ec15ed1035febab4b</summary>

![image](https://github.com/user-attachments/assets/d029b5ab-0225-4678-a0d6-0e92a05669e9)

</details>

jadijadi commented:
> * Network addresses that your Bitcoin node is currently using to communicate with other nodes.

Thanks @pablomartin4btc . Added this as the tooltip & thanks for the tACK

furszy reviewed (commented):
utACK 189c987
